### PR TITLE
Fix the RNN contiguous warning if the whole model parameters are flattened.

### PR DESCRIPTION
**Pull request by alpha0422**
Users may want to flatten the parameters of whole model to achieve high performance. For example, some of the MLPerf models submitted are using flattened parameters to reduce the number of CUDA kernels launched, especially there are tons of small parameters.

However, if the RNN parameters are copied to a new place, it will throw a lot of RNN contiguous warnings, although the RNN parameters are *flattened* at new place too.

This PR try to fix the problem by detecting the weight_buf with an offset, which will return the flattened weight_buf correctly and get rid of the RNN contiguous warnings.

As a simple example, please run following codes:

```python
#!/usr/bin/env python

import torch

torch.manual_seed(1234)
model = torch.nn.Sequential(
    torch.nn.Embedding(32320, 1024),
    torch.nn.LSTM(1024, 1024),
).cuda().half()

nelem = 0
for p in model.parameters():
    nelem += p.numel()

params = torch.empty(nelem, dtype=torch.half, device='cuda')
with torch.no_grad():
    ptr = 0
    for p in model.parameters():
        params[ptr:ptr+p.numel()].copy_(p.data.view(-1))
        p.set_(source=params.storage(), storage_offset=ptr, size=p.size())
        ptr += p.numel()

x = torch.randint(0, 32320, (128, 38), device='cuda')
y = model(x)

torch.cuda.synchronize()
print(y[0].sum())
```

### CODE REVIEW

**Comment by gchanan** [comment]
@zou3519 can you take a look?

off the top of my head:
1) I'm not sure that flattening all the model parameters would give you better performance (do you have a benchmark?)
2) the process is clearly invalid, as the model parameters can be sparse, a different dtype, etc. that wouldn't let you flatten.

**Comment by alpha0422** [comment]
@gchanan 

1. We consider flattening mainly due to two aspects, both of them related to optimizer:

- It reduces the number of kernels launched for optimizer, IIRC Adam launches 7 kernels for each parameter, thus once the parameters/gradients are flattened, you may improve performance if you have tons of relatively small parameters;

- We can distribute the optimizer work to multiple GPUs, thus the optimizer time improved by a huge extent, and the easiest way to implement a distributed optimizer is to make the parameters/gradients flattened by nature (although for now make gradients flattened by nature is not easy in current PyTorch);

Flattening is not the only way to do both of them but is the easiest way with minimal changes. That's why we try to get rid of the RNN warnings although the weight_buf is copied and flattened.

2. We were considering dense tensor, with the same shape. Sparse tensor might be problem, but different type (float, half, etc.) might not have too much influence on flattening itself. However, I think it's details of how to implement flattening, may not affect us to check the RNN weight_buf of this PR, right?

File: aten/src/ATen/native/cudnn/RNN.cpp [hunk 1]
```diff
@@ -1191,7 +1191,10 @@ Tensor try_get_weight_buf(
   // Try to get parameter storage
   auto & any_param = parameters.at(0);
   auto param_storage = any_param.storage();
-  auto weight_buf = at::empty({0}, any_param.options()).set_(param_storage);
+  auto param_offset = any_param.storage_offset();
+  std::initializer_list<int64_t> buffer_size = {
```
Comment (gchanan) [inline aten/src/ATen/native/cudnn/RNN.cpp]: why is this 2-dimensional?  Wasn't it previously 1-dimensional?

Comment (alpha0422) [reply aten/src/ATen/native/cudnn/RNN.cpp]: you're right, it should be 1-dim

**Comment by alpha0422** [comment]
Apart from the performance gains from optimizer, the RNN warning itself indicates the additional copy is a less efficient behavior, so I'd like to say for our cases we don't want to see any RNN warnings at all.

Regarding the uncertainty of whether should support the use case (flatten whole model parameters), is there any specific reason for that? I believe distributed optimizer is one trend as the optimizer is more and more complex, flatten can help distribute the work to multiple workers.

**Comment by Evpok** [comment]
Any new on this @gqchen , @zou3519 , @alpha0422  ?

**Comment by alpha0422** [comment]
We still think it's important to support the situation where users flatten the whole model's parameters into a big chunk for better performance (part of the benefit comes from kernel launch overhead). The PR here want to fix the performance issue with RNN, where if RNN parameters are not located at the start of its storage, then PyTorch deem the RNN parameters as not contiguous and flatten the RNN parameters every iteration. So this PR just changing the RNN contiguous detection logic by introducing the storage offset of the first RNN parameter. With the simple fix, RNN parameters will not be flattened again. It's important to some of the high performance applications.

@zou3519 @gchanan Could you take a look again and see if there's any issues?

**Comment by csarofeen** [comment]
@gchanan @zou3519 this check looks strictly more correct to me, are there remaining concerns with it?

**Review by gchanan — APPROVED** [review]

**Comment by gchanan** [comment]
@csarofeen: I agree it's better.  We aren't agreeing to keep supporting this use case going forward, however.

**Comment by csarofeen** [comment]
@gchanan understood, thanks!
